d2renderers/d2svg/d2svg.go

```diff
@@ -494,6 +495,185 @@ func makeLabelMask(labelTL *geo.Point, width, height int, opacity float64) strin
+
+	switch *pathType {
+	case "M":
+		increment = 3
+	case "L":
+		increment = 3
+	case "C":
+		increment = 7
+	case "S":
+		increment = 5
+	default:
+		return 0, fmt.Errorf("unknown svg path command \"%s\"", *pathType)
+	}
+
+	return increment, nil
+}
+
+// This function finds the length of a path in SVG notation
+func pathLength(pathData []string) (float64, error) {
+	var x, y, pathLength float64
+	var prevPosition geo.Point
+	var increment int
+
+	for i := 0; i < len(pathData); i += increment {
+		switch pathData[i] {
+		case "M":
+			x, _ = strconv.ParseFloat(pathData[i+1], 64)
+			y, _ = strconv.ParseFloat(pathData[i+2], 64)
+		case "L":
+			x, _ = strconv.ParseFloat(pathData[i+1], 64)
+			y, _ = strconv.ParseFloat(pathData[i+2], 64)
+
+			pathLength += geo.EuclideanDistance(prevPosition.X, prevPosition.Y, x, y)
+		case "C":
+			x, _ = strconv.ParseFloat(pathData[i+5], 64)
+			y, _ = strconv.ParseFloat(pathData[i+6], 64)
+
+			pathLength += geo.EuclideanDistance(prevPosition.X, prevPosition.Y, x, y)
+		case "S":
+			x, _ = strconv.ParseFloat(pathData[i+3], 64)
+			y, _ = strconv.ParseFloat(pathData[i+4], 64)
+
+			pathLength += geo.EuclideanDistance(prevPosition.X, prevPosition.Y, x, y)
+		default:
+			return 0, fmt.Errorf("unknown svg path command \"%s\"", pathData[i])
+		}
+
+		prevPosition = geo.Point{X: x, Y: y}
+
+		incr, err := getPathStringIncrement(&pathData[i])
+
+		if err != nil {
+			return 0, err
+		}
+
+		increment = incr
+	}
+
+	return pathLength, nil
+}
+
+// Splits an SVG path into two SVG paths, with the first path being ~{percentage}% of the path
+func splitPath(path string, percentage float64) (string, string, error) {
+	var sumPathLens, curPathLen, x, y float64
+	var prevPosition geo.Point
+	var path1, path2 string
+	var increment int
+
+	pastHalf := false
+	pathData := strings.Split(path, " ")
+	pathLen, err := pathLength(pathData)
+
+	if err != nil {
+		return "", "", err
+	}
+
+	for i := 0; i < len(pathData); i += increment {
+
+		switch pathData[i] {
+		case "M":
+			x, _ = strconv.ParseFloat(pathData[i+1], 64)
+			y, _ = strconv.ParseFloat(pathData[i+2], 64)
+		case "L":
+			x, _ = strconv.ParseFloat(pathData[i+1], 64)
+			y, _ = strconv.ParseFloat(pathData[i+2], 64)
+
+			curPathLen = geo.EuclideanDistance(prevPosition.X, prevPosition.Y, x, y)
+		case "C":
+			x, _ = strconv.ParseFloat(pathData[i+5], 64)
+			y, _ = strconv.ParseFloat(pathData[i+6], 64)
+
+			curPathLen = geo.EuclideanDistance(prevPosition.X, prevPosition.Y, x, y)
+		case "S":
+			x, _ = strconv.ParseFloat(pathData[i+3], 64)
+			y, _ = strconv.ParseFloat(pathData[i+4], 64)
+
+			curPathLen = geo.EuclideanDistance(prevPosition.X, prevPosition.Y, x, y)
+		default:
+			return "", "", fmt.Errorf("unknown svg path command \"%s\"", pathData[i])
+		}
+
+		curPath, err := getSVGPathString(&pathData[i], i, pathData)
+		if err != nil {
+			return "", "", err
+		}
+
+		if pastHalf {
+			path2 += curPath
+		} else if sumPathLens+curPathLen < pathLen*percentage {
```

Comment:
is it possible to not be either these? i thought it's either pastHalf or not pastHalf?